### PR TITLE
Add support for iOS build targets

### PR DESCRIPTION
**Pull request by dbrgn**
Previously compiling for iOS targets (e.g. using [cargo-lipo](https://github.com/TimNN/cargo-lipo)) would fail.

This PR adds support for iOS cross-compile targets, roughly based on [this script](https://github.com/jedisct1/libsodium/blob/master/dist-build/ios.sh) from the libsodium sources.

An Xcode installation on macOS is required for this to work.

### CODE REVIEW

File: rust_sodium-sys/build.rs [hunk 1]
```diff
@@ -394,3 +557,14 @@ fn main() {
...
+
+#[cfg(all(not(windows), not(feature = "use-installed-libsodium")))]
+fn main() {
+    use std::env;
+    let target = unwrap!(env::var("TARGET"));
+    if target.contains("-ios") {
+        main_ios();
+    } else {
+        main_non_ios();
+    }
```
Comment (dbrgn) [inline rust_sodium-sys/build.rs:569]: I tried to do this check with conditional compilation (`#[cfg(target_family = "ios")]`), but for some reason that did not seem to work... And maybe it's better if the code is compiled in unconditionally (on non-windows at least), to be able to catch compile errors even when not compiling for an iOS target.

**Comment by dbrgn** [comment]
Hm, I'm not sure if the test failure is related to my changes...

**Comment by Fraser999** [comment]
Hey @dbrgn,

Thanks very much for this PR - I'm going to get a look at this as soon as possible and will come back here with any comments.

**Comment by Fraser999** [comment]
@dbrgn This is nice work - thanks!

I've actually based a new PR (#48) off this.  Something I'd appreciate your thoughts on is the inclusion of a couple of compiler flags which are in the libsodium shell script for iOS.  These are [`-fembed-bitcode`](https://github.com/Fraser999/rust_sodium/blob/60685b01c1e0693ba36b7343ee571c3820f1de1f/rust_sodium-sys/build.rs#L382) and [`-mthumb`](https://github.com/Fraser999/rust_sodium/blob/60685b01c1e0693ba36b7343ee571c3820f1de1f/rust_sodium-sys/build.rs#L389).  Having briefly read about them, their inclusion make sense to me, and I guess I'd like to keep our script as close as possible to the original.

**Comment by dbrgn** [comment]
@Fraser999 I'm not really familiar with those options but also researched a bit.

It looks like [-fembed-bitcode](https://www.infoq.com/articles/ios-9-bitcode) is something that gives more power to Apple, but can also be used for optimizations. If libsodium includes the option by default, we could probably do so too.

Regarding `mthumb`, i found [this](http://wanderingcoder.net/2010/07/19/ought-arm/):

> When targeting ARMv6, compiling for Thumb is a big tradeoff. ARMv6 Thumb code has access to fewer registers, does not have conditional instructions, and in particular cannot use the floating-point hardware. This means for every single floating-point addition, subtraction, multiplication, etc., floating-point Thumb code must call a system function to do it. Yes, this is as slow as it sounds. For this reason, I recommend disabling Thumb mode when targeting ARMv6. If you do leave it on, make sure you profile your code, and if some parts are slow you should first try disabling Thumb at least for that part (easy with file-specific compiler flags in Xcode, use -mno-thumb). Remember that floating-point calculations are pretty common on iOS since Quartz and Core Animation use a floating-point coordinate system.
> 
> When targeting ARMv7, however, all these drawbacks disappear: ARMv7 contains Thumb-2, an extension of the Thumb instruction set which adds support for conditional execution and 32-bit Thumb instructions that allow access to all ARM registers as well as hardware floating-point and NEON. It’s pretty much a free reduction in code size, so it should be left on (or reenabled if you disabled it); use conditional build settings in Xcode to have it enabled for ARMv7 but disabled for ARMv6.

Since we're only targeting ARMv7 and newer, I think it's good to add/keep the option.

**Comment by Fraser999** [comment]
That's great, thanks.  Ties in with what I found out too.

**Pull request merged**